### ISSUE TITLE
Signature failure after decrypting assertion

### ISSUE THREAD

**Issue by rastko-florencehc** [tngan/samlify#307]
I'm using version 2.5.0.

I'm trying to setup integration with SalesForce with SalesForce being an IdP. When setting up the integration to do only assertion signing everything is working properly. But when I setup the encryption I get the error "ERR_FAILED_TO_VERIFY_SIGNATURE". Decryption goes well but afterwards signature verification fail, digests are not matching to be more precise.

- [this is the signing only response](https://gist.github.com/rastko-florencehc/c679afe896f974657d444d2b844614c6)
- [this is the decrypted response](https://gist.github.com/rastko-florencehc/de80864ad11c4119e5ee115f9038a4a9)

I can see there is a difference in attributes which I think is causing the digest verification to fail. My question is, is this "lost" in the decryption?

**Comment by tngan** [comment]
The decrypted response is using `urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified` but signing only response is taking `urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress`.

However, I think this is not the cause, samlify doesn't manipulate the response, the one you decrypted is the raw xml for sure.

**Comment by tngan** [comment]
It might be caused by

1) using wrong key to verify the signature
2) need to dig dipper to see if the flow has problem

Please double check 1 first, and I might need to take some time to check 2.

**Comment by rastko-florencehc** [comment]
Yes nameid format is not causing this error. I've doubled checked the 1. and proper keys are being used.

Another observation:

If I use [SAML online tool](https://www.samltool.com/online_tools.php) to sign and then encrypt SAML response. And then I tried to use that SAML response (in my integration tests) to decrypt it and verify signature I'm seeing the same behavior - decryption goes well but the signature fails. In this case I've narrowed it down (again) to digests not matching. Furthermore digests are not matching because after decryption, element `<Assertion>` got an extra attribute `xmlns:samlp=""`.  I've sent you the details of this use case via email ~10 days ago.

**Comment by sourcec0de** [comment]
I'm running to the same issue as @rastko-florencehc

**Comment by benjaminlaib** [comment]
I am getting the same error. I have done some debugging, and I noticed, that `verifySignature` in `libsaml.ts` is called with the decrypted document, however it will try to verify it using the original document's signature. Of course this will not work, as the signature is only valid for the original document.

**Comment by benjaminlaib** [comment]
@rastko-florencehc I think I solved it after a lot more debugging!

The library works as expected. There is an IdP option called `messageSigningOrder`. It can be either `'encrypt-then-sign'` or `'sign-then-encrypt'` and defaults to the latter, however the IdP I was using encrypted the response first and signed it afterwards.

Btw, you can use the `ETS` and `STE` properties on `samlify.Constants.MessageSignatureOrder` to avoid hard-coded strings :)

**Comment by tngan** [comment]
@benjaminlaib For the signature, is it assertion signature or message signature? I need to review and add back missing unit tests to take care of all common use cases.

**Comment by benjaminlaib** [comment]
As far as I remember, the IdP I'm using signed both the message and the assertion, but I am not sure. I am not in the office anymore, but I can send you a gist with the SAML documents I'm sending and receiving tomorrow.

However you might be able to have a look for yourself, I am using samlidp.io for testing, that's the service that generated the SAMLResponse in question. It's free and quick to set up.

**Comment by benjaminlaib** [comment]
@tngan I just looked it up and the SAMLResponse actually does have both a message signature as well as an assertion signature.
